ops/relation.py

```diff
@@ -0,0 +1,548 @@
+    done only if the provider charm is ready to service requests.
+    """
+    pass
+
+
+class ProviderBroken(EventBase):
+    """Event raised when provider consumer relation is disolved
+
+    If the realtion between a provider and consumer charm is removed,
+    then a :class:`ProviderBroken` relation is raised.
+    """
+    pass
+
+
+class ConsumerEvents(CharmEvents):
+    """Descriptor for consumer charm events
+    """
+    available = EventSource(ProviderAvailable)
+    invalid = EventSource(ProviderInvalid)
+    unready = EventSource(ProviderUnready)
+    broken = EventSource(ProviderBroken)
+
+
+class Consumer(Object):
+    """Manages relations with a resource provider
+
+    The :class:`Consumer` object manages relations with resource
+    provider charms, by checking compatibility between consumer
+    requirements and provider type and version specification. Any
+    charm that uses resources provided by other charms may manage its
+    relation with the providers by instantiating on :class:`Consumer`
+    object for each such relation. A :class:`Consumer` object may be
+    instantiated in the `__init__` method of the consumer charm as
+    follows
+
+    Example::
+
+        self.provider_name = Consumer(self, resource_name, consumes)
+
+    In managing the relation between provider and consumer, the
+    :class:`Consumer` object may raise any of the following events,
+    that a consumer charm can choose to respond to
+
+    - :class:`ProviderAvailable`
+    - :class:`ProviderInvalid`
+    - :class:`ProviderUnready`
+    - :class:`ProviderBroken`
+
+    Args:
+        charm: :class:`ops.charm.CharmBase` derived object that is
+            instantiating the :class:`Consumer` object. This is almost
+            always `self`.
+        name: string name of relation (as defined in `metadata.yaml`) that
+            consumer charms will use to interact with provider charms.
+        consumes: a dictionary containing acceptable provider
+            specifications. The dictonary must contain at least one
+            key `consumes`. The value of this key is itself a
+            dictionary. This nested dictionary contains key value
+            pairs that are all the acceptable provider
+            specifications. The keys in these specifications are the
+            resource names strings. And the values are version
+            specification strings. The version specification strings
+            can by in any form that is compatible with the
+            [semver](https://pypi.org/project/semver/) Python
+            package. A valid example of the `consumes` dictionary is
+            Example::
+
+            {'consumes': {'mysql': '>5.0.2', 'mariadb': '<=6.1.0'}}
+        multi: a boolean flag that indicates if the :class:`Consumer` object
+            should allow multiple relations with the same relation name. By
+            default this is `False`.
+
+    """
+    on = ConsumerEvents()
+    stored = StoredState()
```

Comment:
similar comments about public visibility of stored, on_foo_changed, etc.

Note that 'on' *is* part of the expected API, since you do expect people to do:

`self.framework.observe(self.consumer.on.available, ...)`

Reply:
Agreed will add underscore prefix to stored.

Reply:
Done.